modules/tensor_mechanics/src/materials/ComputeElasticSmearedCrackingStress.C

```diff
@@ -0,0 +1,634 @@
+    // 4.  Update the rotation tensor to reflect the effect of the 2 eigenvectors.
+
+    // 1.
+    RankTwoTensor R( (*_crack_rotation)[_qp]);
+    RankTwoTensor ePrime = R.transpose() * _elastic_strain[_qp] * R ; //elastic_strain tensor in principial coordinate
+
+    // 2.
+    ColumnMajorMatrix e2x2(2,2);
+    e2x2(0,0) = ePrime(0, 0);
+    e2x2(1,0) = ePrime(1, 0);
+    e2x2(0,1) = ePrime(0, 1);
+    e2x2(1,1) = ePrime(1, 1);
+
+    // 3.
+    ColumnMajorMatrix e_val2x1(2, 1);
+    ColumnMajorMatrix e_vec2x2(2, 2);
+    e2x2.eigen(e_val2x1, e_vec2x2);
+
+    // 4.
+    ColumnMajorMatrix e_vec(3,3);
+    e_vec(0,0) = e_vec2x2(0, 0);
+    e_vec(1,0) = e_vec2x2(1, 0);
+    e_vec(2,0) = 0;
+    e_vec(0,1) = e_vec2x2(0, 1);
+    e_vec(1,1) = e_vec2x2(1, 1);
+    e_vec(2,1) = 0;
+    e_vec(2,0) = 0;
+    e_vec(2,1) = 0;
+    e_vec(2,2) = 1;
+
+    RankTwoTensor eigvec(e_vec(0, 0), e_vec(1, 0), e_vec(2, 0),
+                         e_vec(0, 1), e_vec(1, 1), e_vec(2, 1),
+                         e_vec(0, 2), e_vec(1, 2), e_vec(2, 2));
+
+    (*_crack_rotation)[_qp] = (*_crack_rotation_old)[_qp] * eigvec; //Roe implementation
+
+    principal_strain(0, 0) = e_val2x1(0, 0);
+    principal_strain(1, 0) = e_val2x1(1, 0);
+    principal_strain(2, 0) = ePrime(2, 2);
+  }
+  else if (numKnownDirs == 2 ||
+           numKnownDirs == 3)
+  {
+    // Rotate to cracked orientation and pick off the strains in the rotated
+    // coordinate directions.
+    RankTwoTensor R((*_crack_rotation)[_qp]);
+    RankTwoTensor ePrime = R.transpose() * _elastic_strain[_qp] * R;
+
+    principal_strain(0,0) = ePrime(0, 0);
+    principal_strain(1,0) = ePrime(1, 1);
+    principal_strain(2,0) = ePrime(2, 2);
+  }
+  else
+    mooseError("Invalid number of known crack directions");
+}
+
+unsigned int
+ComputeElasticSmearedCrackingStress::getNumKnownCrackDirs() const
+{
+  const unsigned fromElement(0);
+  unsigned int retVal(0);
+  for (unsigned int i(0); i < 3-fromElement; ++i)
+    retVal += ((*_crack_flags_old)[_qp](i) < 1);
+  return retVal + fromElement;
+}
+
+Real
+ComputeElasticSmearedCrackingStress::computeCrackFactor( int i, Real & sigma, Real & flagVal )
+{
+  if (_cracking_release == CR_EXPONENTIAL)
+  {
+    if ((*_crack_max_strain)[_qp](i) < (*_crack_strain)[_qp](i))
+    {
+      std::stringstream err;
+      err << "Max strain less than crack strain: " << i << " " << sigma << ", "
+          << (*_crack_max_strain)[_qp](i) << ", "
+          << (*_crack_strain)[_qp](i) << ", "
+          << _principal_strain(0, 0) << ", "
+          << _principal_strain(1, 0) << ", "
+          << _principal_strain(2, 0) <<  ", "
+          << _elastic_strain[_qp](0, 0) <<  ", "
+          << _elastic_strain[_qp](1, 1) <<  ", "
+          << _elastic_strain[_qp](2, 2) <<  ", "
+          << _elastic_strain[_qp](0, 1) <<  ", "
+          << _elastic_strain[_qp](0, 2) <<  ", "
+          << _elastic_strain[_qp](1, 2)
+          << std::endl;
+      mooseError( err.str() );
```

Comment:
Building a stringstream here is unnecessary. You can use `<<` inside the `mooseError` macro